diff --git a/drivers/clock_control/clock_stm32l4_l5_wb_wl.c b/drivers/clock_control/clock_stm32l4_l5_wb_wl.c
index 8ddb1e50..8e502a5b 100644
--- a/drivers/clock_control/clock_stm32l4_l5_wb_wl.c
+++ b/drivers/clock_control/clock_stm32l4_l5_wb_wl.c
@@ -8,6 +8,7 @@
 
 #include <soc.h>
 #include <stm32_ll_bus.h>
+#include <stm32_ll_crs.h>
 #include <stm32_ll_pwr.h>
 #include <stm32_ll_rcc.h>
 #include <stm32_ll_utils.h>
@@ -100,4 +101,11 @@ void config_enable_default_clocks(void)
 	/* HW semaphore Clock enable */
 	LL_AHB3_GRP1_EnableClock(LL_AHB3_GRP1_PERIPH_HSEM);
 #endif
+
+#if defined(CRS)
+	/* see https://github.com/zephyrproject-rtos/zephyr/issues/53937 */
+	LL_APB1_GRP1_EnableClock(LL_APB1_GRP1_PERIPH_CRS);
+	LL_CRS_EnableAutoTrimming();
+	LL_CRS_EnableFreqErrorCounter();
+#endif /* defined(CRS) */
 }
