diff --git a/app/src/ble.c b/app/src/ble.c
index e63c63b..8892ded 100644
--- a/app/src/ble.c
+++ b/app/src/ble.c
@@ -220,6 +220,7 @@ static void clear_profile_bond(uint8_t profile) {
 void zmk_ble_clear_bonds(void) {
     LOG_DBG("zmk_ble_clear_bonds()");
 
+    zmk_ble_prof_disconnect(active_profile);
     clear_profile_bond(active_profile);
     update_advertising();
 };
@@ -229,6 +230,7 @@ void zmk_ble_clear_all_bonds(void) {
 
     // Unpair all profiles
     for (int i = 0; i < ZMK_BLE_PROFILE_COUNT; i++) {
+        zmk_ble_prof_disconnect(i);
         clear_profile_bond(i);
     }
 
