diff --git a/quantum/vial.c b/quantum/vial.c
index 93bb15e4e9..7b4d74e017 100644
--- a/quantum/vial.c
+++ b/quantum/vial.c
@@ -340,13 +340,11 @@ static uint8_t dance_state[VIAL_TAP_DANCE_ENTRIES];
 static vial_tap_dance_entry_t td_entry;
 
 static uint8_t dance_step(qk_tap_dance_state_t *state) {
+    /* FIX_VIAL_TAP_DANCE_BEHAVIOR (serach key for checking patch applied) */
     if (state->count == 1) {
-        if (state->interrupted || !state->pressed) return SINGLE_TAP;
-        else return SINGLE_HOLD;
+        return state->pressed ? SINGLE_HOLD : SINGLE_TAP;
     } else if (state->count == 2) {
-        if (state->interrupted) return DOUBLE_SINGLE_TAP;
-        else if (state->pressed) return DOUBLE_HOLD;
-        else return DOUBLE_TAP;
+        return state->pressed ? DOUBLE_HOLD : DOUBLE_TAP;
     }
     return MORE_TAPS;
 }
