diff --git a/quantum/process_keycode/process_tap_dance.c b/quantum/process_keycode/process_tap_dance.c
index 11df6276..b1945795 100644
--- a/quantum/process_keycode/process_tap_dance.c
+++ b/quantum/process_keycode/process_tap_dance.c
@@ -132,6 +132,10 @@ bool preprocess_tap_dance(uint16_t keycode, keyrecord_t *record) {
 
     if (!record->event.pressed) return false;
 
+#ifdef ACTION_FOR_KEYCODE_ENABLE
+    if (record->keycode) return false;
+#endif
+
     if (!active_td || keycode == active_td) return false;
 
     action                             = tap_dance_get(QK_TAP_DANCE_GET_INDEX(active_td));
