diff --git a/quantum/keyboard.c b/quantum/keyboard.c
index df1dc1c3..5410d124 100644
--- a/quantum/keyboard.c
+++ b/quantum/keyboard.c
@@ -520,6 +520,10 @@ static inline void generate_tick_event(void) {
     }
 }
 
+#ifdef CUSTOM_MATRIX_TASK
+extern bool custom_matrix_task(void);
+#    define matrix_task() custom_matrix_task()
+#else
 /**
  * @brief This task scans the keyboards matrix and processes any key presses
  * that occur.
@@ -581,6 +585,7 @@ static bool matrix_task(void) {
 
     return matrix_changed;
 }
+#endif
 
 /** \brief Tasks previously located in matrix_scan_quantum
  *
