diff --git a/tmk_core/protocol/report.h b/tmk_core/protocol/report.h
index 9053b0bb..2e67ad89 100644
--- a/tmk_core/protocol/report.h
+++ b/tmk_core/protocol/report.h
@@ -134,7 +134,8 @@ enum desktop_usages {
 
 // clang-format on
 
-#define NKRO_REPORT_BITS 30
+// default 30, I never use usege id greater than LANG2
+#define NKRO_REPORT_BITS 19
 
 #ifdef KEYBOARD_SHARED_EP
 #    define KEYBOARD_REPORT_SIZE 9
