diff --git a/quantum/encoder.c b/quantum/encoder.c
index 3aee340249..07cad3c7e5 100644
--- a/quantum/encoder.c
+++ b/quantum/encoder.c
@@ -52,7 +52,10 @@ static uint8_t encoder_resolutions[NUM_ENCODERS] = ENCODER_RESOLUTIONS;
 #    define ENCODER_CLOCKWISE false
 #    define ENCODER_COUNTER_CLOCKWISE true
 #endif
-static int8_t encoder_LUT[] = {0, -1, 1, 0, 1, 0, 0, -1, -1, 0, 0, 1, 0, 1, -1, 0};
+#ifndef ENCODER_LOOKUP_TABLE
+#    define ENCODER_LOOKUP_TABLE {0, -1, 1, 0, 1, 0, 0, -1, -1, 0, 0, 1, 0, 1, -1, 0}
+#endif
+static int8_t encoder_LUT[] = ENCODER_LOOKUP_TABLE;
 
 static uint8_t encoder_state[NUM_ENCODERS]  = {0};
 static int8_t  encoder_pulses[NUM_ENCODERS] = {0};
