diff --git a/builddefs/common_rules.mk b/builddefs/common_rules.mk
index cfd261737c..9775922deb 100644
--- a/builddefs/common_rules.mk
+++ b/builddefs/common_rules.mk
@@ -50,7 +50,7 @@ ifeq ($(strip $(LTO_ENABLE)), yes)
         $(info Enabling LTO on arm_atsam-targeting boards is known to have a high likelihood of failure.)
         $(info If unsure, set LTO_ENABLE = no.)
     endif
-    CDEFS += -flto
+    CDEFS += -flto=auto
     CDEFS += -DLTO_ENABLE
 endif
 
