diff --git a/quantum/process_keycode/process_tap_dance.c b/quantum/process_keycode/process_tap_dance.c
index b8a8d32f..3a8a64fe 100644
--- a/quantum/process_keycode/process_tap_dance.c
+++ b/quantum/process_keycode/process_tap_dance.c
@@ -131,6 +131,10 @@ bool preprocess_tap_dance(uint16_t keycode, keyrecord_t *record) {
 
     if (!record->event.pressed) return false;
 
+#ifdef ACTION_FOR_KEYCODE_ENABLE
+    if (record->keycode) return false;
+#endif
+
     if (!active_td || keycode == active_td) return false;
 
     action                             = &tap_dance_actions[TD_INDEX(active_td)];
