diff --git a/quantum/keyboard.c b/quantum/keyboard.c
index e0c9373165..87531f0a31 100644
--- a/quantum/keyboard.c
+++ b/quantum/keyboard.c
@@ -582,6 +582,10 @@ static inline void generate_tick_event(void) {
 
 matrix_row_t matrix_previous[MATRIX_ROWS];
 
+#ifdef CUSTOM_MATRIX_TASK
+extern bool custom_matrix_task(void);
+#    define matrix_task() custom_matrix_task()
+#else
 /**
  * @brief This task scans the keyboards matrix and processes any key presses
  * that occur.
@@ -641,6 +645,7 @@ static bool matrix_task(void) {
 
     return matrix_changed;
 }
+#endif
 
 /** \brief Tasks previously located in matrix_scan_quantum
  *
