diff --git a/subsys/usb/device/usb_device.c b/subsys/usb/device/usb_device.c
index 5baac77c..4fd5d269 100644
--- a/subsys/usb/device/usb_device.c
+++ b/subsys/usb/device/usb_device.c
@@ -151,6 +151,10 @@ struct usb_setup_packet_packed {
 
 static bool reset_endpoint(const struct usb_ep_descriptor *ep_desc);
 
+#if IS_ENABLED(CONFIG_USB_TRACE_SETUP)
+extern void usb_trace_setup(struct usb_setup_packet *setup);
+#endif
+
 /*
  * @brief print the contents of a setup packet
  *
@@ -227,10 +231,16 @@ static bool usb_handle_request(struct usb_setup_packet *setup,
 	}
 
 	if ((*handler)(setup, len, data) < 0) {
+#if IS_ENABLED(CONFIG_USB_TRACE_SETUP)
+		usb_trace_setup(setup);
+#endif
 		LOG_DBG("Handler Error %d", type);
 		usb_print_setup(setup);
 		return false;
 	}
+#if IS_ENABLED(CONFIG_USB_TRACE_SETUP)
+	usb_trace_setup(setup);
+#endif
 
 	return true;
 }
