diff --git a/drivers/adc/adc_nrfx_saadc.c b/drivers/adc/adc_nrfx_saadc.c
index 72a20f47..92ff1092 100644
--- a/drivers/adc/adc_nrfx_saadc.c
+++ b/drivers/adc/adc_nrfx_saadc.c
@@ -322,6 +322,10 @@ static int start_read(const struct device *dev,
 				m_data.positive_inputs[channel_id]);
 			++active_channels;
 		} else {
+			nrf_saadc_burst_set(
+				NRF_SAADC,
+				channel_id,
+				NRF_SAADC_BURST_DISABLED);
 			nrf_saadc_channel_pos_input_set(
 				NRF_SAADC,
 				channel_id,
