diff --git a/app/src/endpoints.c b/app/src/endpoints.c
index ae0e5e7..bae70c7 100644
--- a/app/src/endpoints.c
+++ b/app/src/endpoints.c
@@ -265,6 +265,7 @@ SETTINGS_STATIC_HANDLER_DEFINE(endpoints, "endpoints", NULL, endpoints_handle_se
 
 #endif /* IS_ENABLED(CONFIG_SETTINGS) */
 
+/*
 static bool is_usb_ready(void) {
 #if IS_ENABLED(CONFIG_ZMK_USB)
     return zmk_usb_is_hid_ready();
@@ -280,8 +281,11 @@ static bool is_ble_ready(void) {
     return false;
 #endif
 }
+*/
 
 static enum zmk_transport get_selected_transport(void) {
+    return preferred_transport;
+    /*
     if (is_ble_ready()) {
         if (is_usb_ready()) {
             LOG_DBG("Both endpoint transports are ready. Using %d", preferred_transport);
@@ -299,6 +303,7 @@ static enum zmk_transport get_selected_transport(void) {
 
     LOG_DBG("No endpoint transports are ready.");
     return DEFAULT_TRANSPORT;
+    */
 }
 
 static struct zmk_endpoint_instance get_selected_instance(void) {
