diff --git a/app/module/drivers/display/Kconfig.il0323 b/app/module/drivers/display/Kconfig.il0323
index f3308c1..2704023 100644
--- a/app/module/drivers/display/Kconfig.il0323
+++ b/app/module/drivers/display/Kconfig.il0323
@@ -8,4 +8,12 @@ config IL0323
     depends on SPI
     depends on HEAP_MEM_POOL_SIZE != 0
     help
-      Enable driver for IL0323 compatible controller.
\ No newline at end of file
+      Enable driver for IL0323 compatible controller.
+
+config IL0323_INVERT
+    bool "Invert display"
+    default n
+
+config IL0323_ALTERNATIVE_REFRESH
+    bool "Use an alternative approach for partial refreshes"
+    default n
diff --git a/app/module/drivers/display/il0323.c b/app/module/drivers/display/il0323.c
index c9d72fc..4eef5cf 100644
--- a/app/module/drivers/display/il0323.c
+++ b/app/module/drivers/display/il0323.c
@@ -124,7 +124,11 @@ static int il0323_write(const struct device *dev, const uint16_t x, const uint16
     ptl[IL0323_PTL_HRED_IDX] = x_end_idx;
     ptl[IL0323_PTL_VRST_IDX] = y;
     ptl[IL0323_PTL_VRED_IDX] = y_end_idx;
+#if IS_ENABLED(CONFIG_IL0323_ALTERNATIVE_REFRESH)
+    ptl[sizeof(ptl) - 1] = 0; // limits fading outside of refresh window
+#else
     ptl[sizeof(ptl) - 1] = IL0323_PTL_PT_SCAN;
+#endif
     LOG_HEXDUMP_DBG(ptl, sizeof(ptl), "ptl");
 
     il0323_busy_wait(cfg);
@@ -243,8 +247,12 @@ static void il0323_get_capabilities(const struct device *dev, struct display_cap
     memset(caps, 0, sizeof(struct display_capabilities));
     caps->x_resolution = EPD_PANEL_WIDTH;
     caps->y_resolution = EPD_PANEL_HEIGHT;
-    caps->supported_pixel_formats = PIXEL_FORMAT_MONO10;
+    caps->supported_pixel_formats = PIXEL_FORMAT_MONO10 | PIXEL_FORMAT_MONO01;
+#if IS_ENABLED(CONFIG_IL0323_INVERT)
+    caps->current_pixel_format = PIXEL_FORMAT_MONO01;
+#else
     caps->current_pixel_format = PIXEL_FORMAT_MONO10;
+#endif
     caps->screen_info = SCREEN_INFO_MONO_MSB_FIRST | SCREEN_INFO_EPD;
 }
 
@@ -255,7 +263,7 @@ static int il0323_set_orientation(const struct device *dev,
 }
 
 static int il0323_set_pixel_format(const struct device *dev, const enum display_pixel_format pf) {
-    if (pf == PIXEL_FORMAT_MONO10) {
+    if ((pf == PIXEL_FORMAT_MONO10) || (pf == PIXEL_FORMAT_MONO10)) {
         return 0;
     }
 
