diff --git a/app/src/endpoints.c b/app/src/endpoints.c
index 17cdc5d..d55dd04 100644
--- a/app/src/endpoints.c
+++ b/app/src/endpoints.c
@@ -405,6 +405,7 @@ SETTINGS_STATIC_HANDLER_DEFINE(endpoints, SETTING_SUBTREE, NULL, endpoint_settin
 
 #endif /* IS_ENABLED(CONFIG_SETTINGS) */
 
+/*
 static bool is_usb_ready(void) {
 #if IS_ENABLED(CONFIG_ZMK_USB)
     return zmk_usb_is_hid_ready();
@@ -420,8 +421,10 @@ static bool is_ble_ready(void) {
     return false;
 #endif
 }
-
+*/
 static enum zmk_transport get_selected_transport(void) {
+    return preferred_transport;
+    /*
     switch (preferred_transport) {
     case ZMK_TRANSPORT_NONE:
         LOG_DBG("No endpoint transport selected");
@@ -452,6 +455,7 @@ static enum zmk_transport get_selected_transport(void) {
 
     LOG_DBG("Preferred endpoint transport is %d but no transports are ready", preferred_transport);
     return ZMK_TRANSPORT_NONE;
+    */
 }
 
 static struct zmk_endpoint_instance get_selected_instance(void) {
