diff --git a/app/src/usb_hid.c b/app/src/usb_hid.c
index 945ae989..4c3e1446 100644
--- a/app/src/usb_hid.c
+++ b/app/src/usb_hid.c
@@ -23,6 +23,9 @@
 #endif // IS_ENABLED(CONFIG_ZMK_HID_INDICATORS)
 
 #include <zmk/event_manager.h>
+#if IS_ENABLED(CONFIG_ZMK_USB_HOST_OS)
+#include <zmk/usb_host_os.h>
+#endif // IS_ENABLED(CONFIG_ZMK_USB_HOST_OS)
 
 LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
 
@@ -62,6 +65,11 @@ static uint8_t *get_keyboard_report(size_t *len) {
 
 static int get_report_cb(const struct device *dev, struct usb_setup_packet *setup, int32_t *len,
                          uint8_t **data) {
+
+#if IS_ENABLED(CONFIG_ZMK_USB_HOST_OS)
+    zmk_usb_host_os_trace_hid_setup(setup);
+#endif // IS_ENABLED(CONFIG_ZMK_USB_HOST_OS)
+
     switch (setup->wValue & HID_GET_REPORT_TYPE_MASK) {
     case HID_REPORT_TYPE_FEATURE:
         switch (setup->wValue & HID_GET_REPORT_ID_MASK) {
@@ -119,6 +127,11 @@ static int get_report_cb(const struct device *dev, struct usb_setup_packet *setu
 
 static int set_report_cb(const struct device *dev, struct usb_setup_packet *setup, int32_t *len,
                          uint8_t **data) {
+
+#if IS_ENABLED(CONFIG_ZMK_USB_HOST_OS)
+    zmk_usb_host_os_trace_hid_setup(setup);
+#endif // IS_ENABLED(CONFIG_ZMK_USB_HOST_OS)
+
     switch (setup->wValue & HID_GET_REPORT_TYPE_MASK) {
     case HID_REPORT_TYPE_FEATURE:
         switch (setup->wValue & HID_GET_REPORT_ID_MASK) {
