diff --git a/app/module/drivers/kscan/kscan_gpio_matrix.c b/app/module/drivers/kscan/kscan_gpio_matrix.c
index 8a3c39f..f46260a 100644
--- a/app/module/drivers/kscan/kscan_gpio_matrix.c
+++ b/app/module/drivers/kscan/kscan_gpio_matrix.c
@@ -57,6 +57,9 @@ LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
 #define KSCAN_GPIO_COL_CFG_INIT(idx, inst_idx)                                                     \
     KSCAN_GPIO_GET_BY_IDX(DT_DRV_INST(inst_idx), col_gpios, idx)
 
+#define INTERRUPT_FLAGS                                                                            \
+    COND_CODE_1(IS_ENABLED(CONFIG_GPIO_STM32), (GPIO_INT_EDGE_TO_ACTIVE), (GPIO_INT_LEVEL_ACTIVE))
+
 enum kscan_diode_direction {
     KSCAN_ROW2COL,
     KSCAN_COL2ROW,
@@ -151,7 +154,7 @@ static int kscan_matrix_interrupt_configure(const struct device *dev, const gpio
 
 #if USE_INTERRUPTS
 static int kscan_matrix_interrupt_enable(const struct device *dev) {
-    int err = kscan_matrix_interrupt_configure(dev, GPIO_INT_LEVEL_ACTIVE);
+    int err = kscan_matrix_interrupt_configure(dev, INTERRUPT_FLAGS);
     if (err) {
         return err;
     }
